subsys/net/lib/mqtt_sn/mqtt_sn_transport_udp.c

```diff
@@ -43,25 +44,83 @@ static int tp_udp_init(struct mqtt_sn_transport *transport)
 {
 	struct mqtt_sn_transport_udp *udp = UDP_TRANSPORT(transport);
 	int err;
+	struct sockaddr addrm;
+	struct ip_mreqn mreqn;
+	int optval;
+	struct net_if *iface;
 
-	udp->sock = zsock_socket(udp->gwaddr.sa_family, SOCK_DGRAM, 0);
+	udp->sock = zsock_socket(udp->bcaddr.sa_family, SOCK_DGRAM, 0);
 	if (udp->sock < 0) {
 		return errno;
 	}
 
 	LOG_DBG("Socket %d", udp->sock);
 
-#ifdef LOG_DBG
-	char ip[30], *out;
+	optval = 1;
+	err = zsock_setsockopt(udp->sock, SOL_SOCKET, SO_REUSEADDR, &optval, sizeof(optval));
+	if (err < 0) {
+		return errno;
+	}
+
+	if (IS_ENABLED(CONFIG_MQTT_SN_LOG_LEVEL_DBG)) {
+		char ip[30], *out;
+		uint16_t port;
+
+		out = get_ip_str((struct sockaddr *)&udp->bcaddr, ip, sizeof(ip));
+		switch (udp->bcaddr.sa_family) {
+		case AF_INET:
+			port = ntohs(((struct sockaddr_in *)&udp->bcaddr)->sin_port);
+			break;
+		case AF_INET6:
+			port = ntohs(((struct sockaddr_in6 *)&udp->bcaddr)->sin6_port);
+			break;
+		default:
+			break;
+		}
+
+		if (out != NULL) {
+			LOG_DBG("Binding to Brodcast IP %s:%u", out, port);
+		}
+	}
+
+	switch (udp->bcaddr.sa_family) {
+	case AF_INET:
+		addrm.sa_family = AF_INET;
+		((struct sockaddr_in *)&addrm)->sin_port =
+			((struct sockaddr_in *)&udp->bcaddr)->sin_port;
+		((struct sockaddr_in *)&addrm)->sin_addr.s_addr = INADDR_ANY;
+		break;
+	case AF_INET6:
+		if (IS_ENABLED(CONFIG_NET_IPV6)) {
+			addrm.sa_family = AF_INET6;
+			((struct sockaddr_in6 *)&addrm)->sin6_port =
+				((struct sockaddr_in6 *)&udp->bcaddr)->sin6_port;
+			memcpy(&((struct sockaddr_in6 *)&addrm)->sin6_addr, &in6addr_any,
+			       sizeof(struct in6_addr));
+			break;
+		}
+	default:
+		LOG_ERR("Unknown AF");
+		return -EINVAL;
+	}
+
+	err = zsock_bind(udp->sock, &addrm, sizeof(addrm));
+	if (err) {
+		LOG_ERR("Error during bind: %d", errno);
+		return errno;
+	}
 
-	out = get_ip_str((struct sockaddr *)&udp->gwaddr, ip, sizeof(ip));
-	if (out != NULL) {
-		LOG_DBG("Connecting to IP %s:%u", out,
-			ntohs(((struct sockaddr_in *)&udp->gwaddr)->sin_port));
+	memcpy(&mreqn.imr_multiaddr, &udp->bcaddr.data[2], sizeof(udp->bcaddr.data) - 2);
+	iface = net_if_get_default();
```

Comment:
This might give wrong interface if there are multiple in the system. Perhaps we could select the one where the gwaddr traffic is routed. You can use `net_if_ipv6_select_src_iface()` for IPv6 and `net_if_ipv4_select_src_iface()` for IPv4 to find out the correct interface.